lib/api/objectDelete.js

```diff
@@ -23,44 +29,132 @@ export default function objectDelete(authInfo, request, log, cb) {
+                    delOptions, log, (err, delResult) => next(err, bucketMD,
+                    objectMD, delResult, deleteInfo));
             }
-            if (!objMD) {
-                return cb(errors.NoSuchKey, corsHeaders);
+            // putting a new delete marker
+            deleteInfo.newDeleteMarker = true;
+            return createAndStoreObject(bucketName, bucketMD,
+                objectKey, objectMD, authInfo, canonicalID, null, request,
+                deleteInfo.newDeleteMarker, null, log, (err, newDelMarkerRes) =>
+                next(err, bucketMD, objectMD, newDelMarkerRes, deleteInfo));
+        },
+    ], (err, bucketMD, objectMD, result, deleteInfo) => {
+        const resHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucketMD);
+        if (err) {
+            log.debug('error processing request', { error: err,
+                method: 'objectDelete' });
+            return cb(err, resHeaders);
+        }
+        if (deleteInfo.newDeleteMarker) {
+            // if we created a new delete marker, return true for
+            // x-amz-delete-marker and the version ID of the new delete marker
+            if (result.versionId) {
+                resHeaders['x-amz-delete-marker'] = true;
+                resHeaders['x-amz-version-id'] = result.versionId === 'null' ?
+                    result.versionId : versionIdUtils.encrypt(result.versionId);
             }
-            if (objMD['content-length']) {
-                log.end().addDefaultFields({
-                    contentLength: objMD['content-length'],
-                });
+            // TODO metric for delete marker
+            pushMetric('putObject', log, { authInfo, bucket: bucketName,
+                newByteLength: 0, oldByteLength: 0 });
+        } else {
+            // otherwise, if deleting a specific version or delete marker,
+            // return the version id in the response headers
+            if (reqVersionId) {
+                resHeaders['x-amz-version-id'] = reqVersionId === 'null' ?
+                    reqVersionId : versionIdUtils.encrypt(reqVersionId);
+                if (deleteInfo.removeDeleteMarker) {
+                    resHeaders['x-amz-delete-marker'] = true;
+                }
             }
-            return services.deleteObject(bucketName, objMD, objectKey, log,
-                err => {
-                    if (err) {
-                        return cb(err, corsHeaders);
-                    }
-                    pushMetric('deleteObject', log, {
-                        authInfo,
-                        bucket: bucketName,
-                        byteLength: objMD['content-length'],
-                        numberOfObjects: 1,
-                    });
-                    return cb(null, corsHeaders);
-                });
-        });
+            pushMetric('deleteObject', log, { authInfo, bucket: bucketName,
```

Comment:
Except for unversioned or versioning-suspended buckets, numberOfObjects deleted should be 0.
For versioned buckets, numberOfObjects should be 0 once we set the delete marker.